rotkehlchen/chain/ethereum/modules/makerdao/sai/decoder.py

```diff
@@ -0,0 +1,505 @@
+        amount_withdrawn = asset_normalized_value(amount=amount_withdrawn_raw, asset=self.sai)
+
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.GENERATE_DEBT and
+                event.balance.amount == amount_withdrawn and
+                event.counterparty == CPT_SAI
+            ):
+                return None, []
+
+        if self.base_tools.is_tracked(withdrawer) is True:
+            event = HistoryBaseEntry(
+                event_identifier=transaction.tx_hash,
+                timestamp=ts_sec_to_ms(transaction.timestamp),
+                sequence_index=self.base_tools.get_next_sequence_counter(),
+                location=Location.BLOCKCHAIN,
+                asset=self.sai,
+                event_type=HistoryEventType.RECEIVE,
+                event_subtype=HistoryEventSubType.GENERATE_DEBT,
+                balance=Balance(amount=amount_withdrawn),
+                location_label=withdrawer,
+                counterparty=CPT_SAI,
+                notes=f'Borrow {amount_withdrawn} {self.sai.symbol} from CDP {cdp_id}',
+            )
+            return event, []
+
+        # sai was actually borrowed, but it was handled by a proxy so create an action item
+        # to transform the "receive" event later
+        action_item = ActionItem(
+            action='transform',
+            sequence_index=tx_log.log_index,
+            from_event_type=HistoryEventType.RECEIVE,
+            from_event_subtype=HistoryEventSubType.NONE,
+            asset=self.sai,
+            amount=amount_withdrawn,
+            to_event_type=HistoryEventType.RECEIVE,
+            to_event_subtype=HistoryEventSubType.GENERATE_DEBT,
+            to_notes=f'Borrow {amount_withdrawn} {self.sai.symbol} from CDP {cdp_id}',
+            to_counterparty=CPT_SAI,
+            extra_data={'cdp_id': cdp_id},
+        )
+        return None, [action_item]
+
+    def _decode_repay_sai_event(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[list[ActionItem]],  # pylint: disable=unused-argument
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        """
+        This method decodes an event of SAI loan repayment.
+
+        An example of such transaction is:
+        https://etherscan.io/tx/0xe964cb12f4bbfa1ba4b6db8464eb3f2d4234ceafb0b5ec5f4a2188b0264bab27
+        """
+        cdp_id = hex_or_bytes_to_int(tx_log.topics[2])
+        depositor = hex_or_bytes_to_address(tx_log.topics[1])
+        amount_paid_raw = hex_or_bytes_to_int(tx_log.topics[3])
+        amount_paid = asset_normalized_value(amount=amount_paid_raw, asset=self.sai)
+
+        for _event in decoded_events:
```

Comment:
why are you using an `_` here. Don't do that. This is not our naming convention and you only do it in this function. Wtf?